### ISSUE TITLE
 [idxd_perf] idxd_perf core dumped

### ISSUE THREAD

**Issue by ye1xchen** [spdk/spdk#2503]
# Bug report

./build/examples/idxd_perf Segmentation fault (core dumped)
## Expected Behavior

idxd_perf test normally.

## Current Behavior
./build/examples/idxd_perf  -q 32 -o 4096 -t 5 -w copy -k
EAL: No free 2048 kB hugepages reported on node 1
EAL: No available 1048576 kB hugepages reported
TELEMETRY: No legacy callbacks, legacy socket not created
Segmentation fault (core dumped)

dmesg | tail -20
[  553.278817] show_signal_msg: 23 callbacks suppressed
[  553.278824] idxd_perf[16354]: segfault at 20 ip 000000000041fb43 sp 00007ffd3f655f80 error 4 in idxd_perf[405000+df000]
[  553.278845] Code: 50 00 00 00 00 49 c7 44 24 38 00 00 00 00 49 89 44 24 40 e8 8f 5f fe ff 8b 43 1c 39 43 14 0f 84 a0 02 00 00 48 8b 03 48 89 df <ff> 50 20 48 89 ef 49 89 44 24 08 8b 43 14 89 c2 83 c0 01 c1 e2 06
[  684.478587] pci 0000:02:00.0: PCI bridge to [bus 03]
[  684.478607] pci 0000:02:00.0:   bridge window [io  0x1000-0x1fff]
[  684.478627] pci 0000:02:00.0:   bridge window [mem 0x94000000-0x950fffff]
[  707.583713] IOMMU: dmar15: Page request without PASID
[  707.583723] dmar15: Invalid page request: 6a080001 2000003e9007
[  707.584140] idxd 0000:6a:01.0: err[0]: 0x0000000300011a2d
[  707.584154] idxd 0000:6a:01.0: err[1]: 0x0000000000000000
[  707.584158] idxd 0000:6a:01.0: err[2]: 0x00002000003e9000
[  707.584161] idxd 0000:6a:01.0: err[3]: 0x0000000000000000
[  707.584171] idxd 0000:6a:01.0: err[0]: 0x0000000300011a2c
[  707.584174] idxd 0000:6a:01.0: err[1]: 0x0000000000000000
[  707.584177] idxd 0000:6a:01.0: err[2]: 0x00002000003e9000
[  707.584180] idxd 0000:6a:01.0: err[3]: 0x0000000000000000
[  823.792529] idxd_perf[37357]: segfault at 20 ip 000000000041fb43 sp 00007ffc109d35f0 error 4 in idxd_perf[405000+df000]
[  823.792557] Code: 50 00 00 00 00 49 c7 44 24 38 00 00 00 00 49 89 44 24 40 e8 8f 5f fe ff 8b 43 1c 39 43 14 0f 84 a0 02 00 00 48 8b 03 48 89 df <ff> 50 20 48 89 ef 49 89 44 24 08 8b 43 14 89 c2 83 c0 01 c1 e2 06
[ 1701.174949] idxd_perf[45155]: segfault at 20 ip 000000000041fb43 sp 00007ffe5d1d7c70 error 4 in idxd_perf[405000+df000]
[ 1701.174978] Code: 50 00 00 00 00 49 c7 44 24 38 00 00 00 00 49 89 44 24 40 e8 8f 5f fe ff 8b 43 1c 39 43 14 0f 84 a0 02 00 00 48 8b 03 48 89 df <ff> 50 20 48 89 ef 49 89 44 24 08 8b 43 14 89 c2 83 c0 01 c1 e2 06

## Steps to Reproduce

1../configure --with-idxd
make -j
2.accel-config disable-wq dsa0/wq0.1
accel-config disable-device dsa0
accel-config config-wq --group-id=0 --mode=dedicated --wq-size=16 --type=user --name="MyApp1" --priority=10 --block-on-fault=1 dsa0/wq0.1
accel-config config-engine dsa0/engine0.1 --group-id=0
accel-config enable-device dsa0
accel-config enable-wq dsa0/wq0.1
3../build/examples/idxd_perf -q 1 -o 4096 -t 5 -w copy -k

## Context (Environment including OS version, SPDK version, etc.)
fedora 32
Latest SPDK master(commit 68184a50306837386008844698fcaf236659041f)

**Comment by Comphix** [comment]
Looking at this issue.

**Comment by Comphix** [comment]
First tried the userspace idxd and there is no issue.

# ./build/examples/idxd_perf -q 1 -o 4096 -t 5 -w copy -r 1
EAL: No free 2048 kB hugepages reported on node 1
EAL: No available 1048576 kB hugepages reported
TELEMETRY: No legacy callbacks, legacy socket not created
idxd device=0xcf9270 is bound on core=0
Notice: we cannot let single worker assign idxd devices
more than 1, you need use -r while starting app to change this value
SPDK Configuration:
Core mask:      0x1

Idxd Perf Configuration:
Workload Type:   copy
Transfer size:   4096 bytes
Queue depth:     1
Allocated depth: 1
Run time:        5 seconds
Verify:          No

Starting thread on core 0

IDXD_ChanID   Core      Transfers      Bandwidth     Failed     Miscompares
------------------------------------------------------------------------
         0    0         995249/s     3887 MiB/s      0           0
=========================================================================
Total:                   995249/s     3887 MiB/s     0           0

**Comment by Comphix** [comment]
Using kernel idxd driver, the impl is not set.

(gdb) p idxd->impl
$2 = (struct spdk_idxd_impl *) 0x0 <<<<<

```
(gdb) r
Starting program: /home/gangcao/20160628/waikiki/spdk/build/examples/idxd_perf -q 1 -o 4096 -t 5 -w copy -k
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
[New Thread 0x7ffff7934700 (LWP 405837)]
EAL: No free 2048 kB hugepages reported on node 1
EAL: No available 1048576 kB hugepages reported
[New Thread 0x7ffff7133700 (LWP 405838)]
TELEMETRY: No legacy callbacks, legacy socket not created

Thread 1 "idxd_perf" received signal SIGSEGV, Segmentation fault.
spdk_idxd_get_channel (idxd=0x8a4d30) at idxd.c:120
120             chan->portal = idxd->impl->portal_get_addr(idxd);
(gdb) p idxd
$1 = (struct spdk_idxd_device *) 0x8a4d30
(gdb) p idxd->impl
$2 = (struct spdk_idxd_impl *) 0x0
(gdb) bt
#0  spdk_idxd_get_channel (idxd=0x8a4d30) at idxd.c:120
#1  0x00000000004073fc in init_idxd_chan_entry (t=0x8a4fe0, idxd=0x8a4d30) at perf.c:873
#2  0x00000000004075f3 in associate_workers_with_idxd_device () at perf.c:926
#3  0x0000000000407843 in main (argc=10, argv=0x7fffffffdc68) at perf.c:1010
```

**Comment by Comphix** [comment]
Looks like in the kernel_idxd_probe() kernel probe operation, the impl is not set.

**Comment by peluse** [comment]
I'l look into it

**Comment by peluse** [comment]
Still works fine for me, see output below.  Can you please provide th output for **each** of your accel_config steps? Also I just noticed WQ size of 16.  Any reason for this change from the directions? I doubt its causing an issue but it's not the value I used.

[peluse@fedora spdk]$ ./build/examples/idxd_perf -q 32 -o 4096 -t 5 -w copy -k
EAL: No free 2048 kB hugepages reported on node 1
EAL: No available 1048576 kB hugepages reported
TELEMETRY: No legacy callbacks, legacy socket not created
[2022-05-10 06:54:19.781103] idxd_kernel.c: 114:dsa_setup_single_wq: *ERROR*: Can not open the Working queue file descriptor on path=/dev/char/235:0
[2022-05-10 06:54:19.781127] idxd_kernel.c: 315:_kernel_idxd_probe: *ERROR*: Failed to probe requested wqs on kernel device context=0x2543b50
No idxd device found
[peluse@fedora spdk]$ sudo ./build/examples/idxd_perf -q 32 -o 4096 -t 5 -w copy -k
EAL: No free 2048 kB hugepages reported on node 1
EAL: No available 1048576 kB hugepages reported
TELEMETRY: No legacy callbacks, legacy socket not created
idxd device=0x1e71a90 is bound on core=0
SPDK Configuration:
Core mask:      0x1

Idxd Perf Configuration:
Workload Type:   copy
Transfer size:   4096 bytes
Queue depth:     32
Allocated depth: 32
Run time:        5 seconds
Verify:          No

Starting thread on core 0

IDXD_ChanID   Core      Transfers      Bandwidth     Failed     Miscompares
------------------------------------------------------------------------
         0    0         215605/s      842 MiB/s      0           0
=========================================================================
Total:                   215605/s      842 MiB/s     0           0

**Comment by peluse** [comment]
Still works for me on master using the correct cmd line this time, see below.  Please provide the output for each of the accel_config steps, also why was 16 used for wQ size? Doubt this is the problem but please try using the directions provided on spdk.io

[peluse@fedora examples]$ sudo ./idxd_perf -q 32 -o 4096 -t 5 -w copy -k
EAL: No free 2048 kB hugepages reported on node 1
EAL: No available 1048576 kB hugepages reported
TELEMETRY: No legacy callbacks, legacy socket not created
idxd device=0xfeea90 is bound on core=0
SPDK Configuration:
Core mask:      0x1

Idxd Perf Configuration:
Workload Type:   copy
Transfer size:   4096 bytes
Queue depth:     32
Allocated depth: 32
Run time:        5 seconds
Verify:          No

Starting thread on core 0

IDXD_ChanID   Core      Transfers      Bandwidth     Failed     Miscompares
------------------------------------------------------------------------
         0    0         xxx/s      xxx MiB/s      0           0
=========================================================================
Total:                   xxx/s      xxx MiB/s     0           0

**Comment by Comphix** [comment]
The commit id in my reproduction is:

commit 0849aadceb820b8390902a2f06999c6eced94c6d (HEAD -> bugfix_50, origin/master, origin/HEAD, master)
Author: John Levon <john.levon@nutanix.com>
Date:   Thu May 5 23:34:41 2022 +0100

    nvmf/vfio-user: clarify CQ in handle_queue_connect_rsp()

For the accel_config, below are the steps that I follow Chen Ye's usage:

```
accel-config disable-wq dsa0/wq0.1
accel-config disable-device dsa0
accel-config config-wq --group-id=0 --mode=dedicated --wq-size=16 --type=user --name="MyApp1" --priority=10 --block-on-fault=1 dsa0/wq0.1
accel-config config-engine dsa0/engine0.1 --group-id=0
accel-config enable-device dsa0
accel-config enable-wq dsa0/wq0.1
```

BTW: looks like there is no data output from above running?

**Comment by peluse** [comment]
Hmm, OK I get output, see below. Anyway, can you please test on master and confirm it works for you there too.  I don't know why it doesn't work with this unrelated patch but we really shouldn't be opening issue on patches out for review especially when the issue has nothing to do with the patch.  I'd prefer for this to land first. I did try a quick rebase on master and it still fails but once this patch lands lets revisit this and debug on master OK?

dsa0/wq0.1 is in disabled mode already, skipping...
failed in dsa0/wq0.1
disabled 0 wq(s) out of 1
dsa0 is in disabled state already, skipping...
failed in dsa0
disabled 0 device(s) out of 1
enabled 1 device(s) out of 1
enabled 1 wq(s) out of 1

**Comment by Comphix** [comment]
Sorry for the confusion, above commit id is from the master and I've created my local working Repo here.

Will rebase again to the master and have another testing.

Especially for below config-wq operation, is this all right or any suggested setting?

Thanks.

accel-config config-wq --group-id=0 --mode=dedicated --wq-size=16 --type=user --name="MyApp1" --priority=10 --block-on-fault=1 dsa0/wq0.1

**Comment by Comphix** [comment]
Just rebased to master as following commit id

```
commit 468b3e92555955a0da17aa6a25bf2cab809ed652 (HEAD -> master, origin/master, origin/HEAD)
Author: Boris Glimcher <Boris.Glimcher@emc.com>
Date:   Wed May 4 21:57:37 2022 +0300

    configure: Add correct CFLAGS and LDFLAGS for libssl11
```

It still hit the same issue.

Is there any difference with the accel-config operations or DSA hardware?

**Comment by peluse** [comment]
Thanks, accel_config just configured the work queues so I'm sure it's possible to screw something up but can't say I've tried a bunch of crazy things.  I don't understand why that patch even rebased fails but master works.  Did you try master yet?

**Comment by peluse** [comment]
I'l pull the rebased patch and dig in a bit more too

**Comment by peluse** [comment]
I believe this patch is the culprit but I don't have a fix yet.  @benlwalker its clear where setting of th the kernel impl was removed from _kernel_idxd_probe() but not sure why nothing was added but I haven't gone through it in painstaking detail yet 

https://review.spdk.io/gerrit/c/spdk/spdk/+/11537

**Comment by peluse** [comment]
FYI if you fwd declare the g_kernel_idxd_impl struct and add the following line into kernel_idxd_probe() right after kenrel_idxd struct si allocated it works.  Don't know what else might be missing so lets have Ben take another look at his patch
 
kernel_idxd->idxd.impl = &g_kernel_idxd_impl;

**Comment by Comphix** [comment]
Yes, for this change, had a try yesterday, however it seems that the testing could not exit normally.

Looks like the IO is not really completed so that the CurrentQueueDepth is larger than 0 and the tests cannot exit.

The core dump itself could be fixed by this impl set.

Still need to understand more about this issue.

**Comment by peluse** [comment]
OK, my day is done.  If you get a chance maybe revert the whole patch and see how things run.  I just did that quick 2 liner and it seemed fine

**Comment by optimistyzy** [comment]
> I believe this patch is the culprit but I don't have a fix yet. @benlwalker its clear where setting of th the kernel impl was removed from _kernel_idxd_probe() but not sure why nothing was added but I haven't gone through it in painstaking detail yet
> 
> https://review.spdk.io/gerrit/c/spdk/spdk/+/11537

This patch is the root cause.  As Gang described, idxd->impl is missed setting. 

By the way, in my mind mind, we should not restrict the WQ usage to 1 with this patch.  In idxd.c, when getting the channel in generic layer (idxd), it can be based on WQ. So even if users configure many WQs in the kernel,  the  spdk accel framework can still leverage many WQs well. 

Also without this patch, the test on kernel path works well.

**Comment by peluse** [comment]
Thanks for jumping in @optimistyzy Ziye!!!!

**Comment by peluse** [comment]
Wrong impl @benlwalker  we're not talking about g_idxd_impl we're talking about the pointer to the either user or kernel imply that gets stuffed into the kernel_idxd struct.  See the patch mentioned above and the LOC I mentioned that got removed.

**Comment by benlwalker** [comment]
https://review.spdk.io/gerrit/c/spdk/spdk/+/12641

For me it still isn't working because for some reason my WQs are reporting as kernel model even though I'm setting the type to "user", but this gets a lot further.

**Comment by peluse** [comment]
> https://review.spdk.io/gerrit/c/spdk/spdk/+/12641
> 
> For me it still isn't working because for some reason my WQs are reporting as kernel model even though I'm setting the type to "user", but this gets a lot further.

Yeah, that's the change I suggested above and tested on my system and it worked for me.  @Comphix can you try on your system?

**Comment by Comphix** [comment]
Here it is, as mentioned earlier, in my testing configuration, the core dump is not hit as we set the impl now. However the tests can not complete. Not sure whether it is due to the DSA related configuration. @peluse , could you also share your detailed DSA configuration steps.

```
diff --git a/lib/idxd/idxd_kernel.c b/lib/idxd/idxd_kernel.c
index 6acdd4faf..cdcb1934c 100644
--- a/lib/idxd/idxd_kernel.c
+++ b/lib/idxd/idxd_kernel.c
@@ -62,6 +62,8 @@ struct spdk_kernel_idxd_device {

 #define __kernel_idxd(idxd) SPDK_CONTAINEROF(idxd, struct spdk_kernel_idxd_device, idxd)

+static struct spdk_idxd_impl g_kernel_idxd_impl;
+
 static void
 kernel_idxd_device_destruct(struct spdk_idxd_device *idxd)
 {
@@ -114,6 +116,7 @@ kernel_idxd_probe(void *cb_ctx, spdk_idxd_attach_cb attach_cb)
                kernel_idxd->max_batch_size = accfg_device_get_max_batch_size(device);
                kernel_idxd->max_xfer_size = accfg_device_get_max_transfer_size(device);
                kernel_idxd->idxd.socket_id = accfg_device_get_numa_node(device);
+               kernel_idxd->idxd.impl = &g_kernel_idxd_impl;
                kernel_idxd->fd = -1;

                accfg_wq_foreach(device, wq) {

[root@waikikibeach68 spdk]# ./scripts/setup.sh status
Hugepages
node     hugesize     free /  total
node0   1048576kB        0 /      0
node0      2048kB    13822 /  13824
node1   1048576kB        0 /      0
node1      2048kB        0 /      0

Type     BDF             Vendor Device NUMA    Driver           Device     Block devices
NVMe     0000:27:00.0    8086   0953   0       nvme             nvme0      nvme0n1
IDXD     0000:6a:01.0    8086   0b25   0       idxd             -          -
IDXD     0000:6f:01.0    8086   0b25   0       idxd             -          -
IDXD     0000:74:01.0    8086   0b25   0       idxd             -          -
IDXD     0000:79:01.0    8086   0b25   0       idxd             -          -
IDXD     0000:e7:01.0    8086   0b25   1       idxd             -          -
IDXD     0000:ec:01.0    8086   0b25   1       idxd             -          -
IDXD     0000:f1:01.0    8086   0b25   1       idxd             -          -
IDXD     0000:f6:01.0    8086   0b25   1       idxd             -          -


[root@waikikibeach68 spdk]# accel-config disable-wq dsa0/wq0.1
disabled 1 wq(s) out of 1

[root@waikikibeach68 spdk]# accel-config disable-device dsa0
disabled 1 device(s) out of 1

[root@waikikibeach68 spdk]#  accel-config config-wq --group-id=0 --mode=dedicated --wq-size=16 --type=user --name="MyApp1" --priority=10 --block-on-fault=1 dsa0/wq0.1

[root@waikikibeach68 spdk]# accel-config config-engine dsa0/engine0.1 --group-id=0

[root@waikikibeach68 spdk]# accel-config enable-device dsa0
enabled 1 device(s) out of 1

[root@waikikibeach68 spdk]# accel-config enable-wq dsa0/wq0.1
enabled 1 wq(s) out of 1

[root@waikikibeach68 spdk]# gdb --args ./build/examples/idxd_perf -q 1 -o 4096 -t 5 -w copy -k
GNU gdb (GDB) Fedora 9.1-7.fc32
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./build/examples/idxd_perf...
(gdb) r
Starting program: /home/gangcao/spdk/build/examples/idxd_perf -q 1 -o 4096 -t 5 -w copy -k
Missing separate debuginfos, use: dnf debuginfo-install glibc-2.31-6.fc32.x86_64
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
[New Thread 0x7ffff789d700 (LWP 832456)]
EAL: No free 2048 kB hugepages reported on node 1
EAL: No available 1048576 kB hugepages reported
[New Thread 0x7ffff709c700 (LWP 832457)]
TELEMETRY: No legacy callbacks, legacy socket not created
idxd device=0x8a4d30 is bound on core=0
SPDK Configuration:
Core mask:      0x1

Idxd Perf Configuration:
Workload Type:   copy
Transfer size:   4096 bytes
Queue depth:     1
Allocated depth: 1
Run time:        5 seconds
Verify:          No

Starting thread on core 0
^C
Thread 1 "idxd_perf" received signal SIGINT, Interrupt.
0x0000000000406431 in drain_io (t=0x8a4fe0) at perf.c:510
510             while (t->current_queue_depth > 0) {
Missing separate debuginfos, use: dnf debuginfo-install libuuid-2.35.2-1.fc32.x86_64 numactl-libs-2.0.12-4.fc32.x86_64 openssl-libs-1.1.1k-1.fc32.x86_64 zlib-1.2.11-21.fc32.x86_64
(gdb) bt
#0  0x0000000000406431 in drain_io (t=0x8a4fe0) at perf.c:510
#1  0x000000000040734d in work_fn (arg=0x889800) at perf.c:830
#2  0x0000000000407989 in main (argc=10, argv=0x7fffffffdc78) at perf.c:1037
(gdb) q
A debugging session is active.

        Inferior 1 [process 832452] will be killed.

Quit anyway? (y or n) y

[root@waikikibeach68 spdk]# git log
commit 673575184412d8acc6cb6f1bdb91133bef13c6be (HEAD -> master, origin/master, origin/HEAD)
Author: luzhipeng <luzhipeng@cestc.cn>
Date:   Fri May 6 18:43:46 2022 +0800
```

**Comment by peluse** [comment]
Yeah, I don't get any of those errors at all.  I followed our docs at https://spdk.io/doc/accel_fw.html

**Comment by Comphix** [comment]
Looks like still same issue. Maybe relate to the DSA itself. I will have a further check.

**Comment by peluse** [comment]
OK, I thought you'd mentioned idxd_perf worked w/the SPDK driver (no -k).  You could try that again and also try accel_perf also.  You could also double check (rebuild or install) your kernel driver I guess

**Comment by Comphix** [comment]
Found some issues in kernel...
```
[Wed May 11 23:17:22 2022] idxd dsa0: Active wq 1 on disable wq0.1.
[Wed May 11 23:18:08 2022] IOMMU: dmar15: Page request without PASID
[Wed May 11 23:18:08 2022] dmar15: Invalid page request: 6a080001 2000003e9027
[Wed May 11 23:18:08 2022] idxd 0000:6a:01.0: err[0]: 0x0000000300021a2d
[Wed May 11 23:18:08 2022] idxd 0000:6a:01.0: err[1]: 0x0000000000000000
[Wed May 11 23:18:08 2022] idxd 0000:6a:01.0: err[2]: 0x00002000003e9000
[Wed May 11 23:18:08 2022] idxd 0000:6a:01.0: err[3]: 0x0000000000000000
[Wed May 11 23:18:08 2022] idxd 0000:6a:01.0: err[0]: 0x0000000300021a2c
[Wed May 11 23:18:08 2022] idxd 0000:6a:01.0: err[1]: 0x0000000000000000
[Wed May 11 23:18:08 2022] idxd 0000:6a:01.0: err[2]: 0x00002000003e9000
[Wed May 11 23:18:08 2022] idxd 0000:6a:01.0: err[3]: 0x0000000000000000
```

**Comment by Comphix** [comment]
Checked out previous commits and looks like same issue with the kernel driver. May relate to kernel driver. Will check this issue again later.

**Comment by Comphix** [comment]
> > https://review.spdk.io/gerrit/c/spdk/spdk/+/12641
> > For me it still isn't working because for some reason my WQs are reporting as kernel model even though I'm setting the type to "user", but this gets a lot further.
> 
> Yeah, that's the change I suggested above and tested on my system and it worked for me. @Comphix can you try on your system?

For this change, it is needed.

**Comment by peluse** [comment]
well, technically the core dump is not a 3rd party issue, it was a bug in our code positively identified with a fix posted above.  I think we should close this one to avoid confusion.  If you have an issue with your system I think it'd be better to open a new issue that explains just that behavior without all of this irrelevant history.

Please close if you agree @Comphix

**Comment by Comphix** [comment]
Ok, let's close this one first.
